### PR TITLE
Stack manipulation opcodes - uncover and cover

### PR DESCRIPTION
**Pull request by AyAggarwal**
This PR adds two opcodes.
cover:  "remove top of stack, and place it down the stack such that N elements are above it",
uncover:  "remove the value at depth N in the stack and shift above items down so the Nth deep value is on top of the stack"

### CODE REVIEW

**Comment by AyAggarwal** [comment]
Hello All, the PR build is failing due to "scripts/travis/codegen_verification.sh". Will look into this tomorrow morning (PST)

File: data/transactions/logic/eval.go [hunk 1]
```diff
@@ -1700,6 +1700,40 @@ func opDig(cx *evalContext) {
 	cx.stack = append(cx.stack, sv)
 }
 
+func opCover(cx *evalContext) {
+	depth := int(uint(cx.program[cx.pc+1]))
+	idx := len(cx.stack) - 1 - depth
+	// Need to check stack size explicitly here because checkArgs() doesn't understand dig
```
Comment (jannotti) [inline data/transactions/logic/eval.go:1706]: Cover, not dig

File: data/transactions/logic/eval.go [hunk 2]
```diff
@@ -1700,6 +1700,40 @@ func opDig(cx *evalContext) {
...
+	// so we can't expect our stack to be prechecked.
+	if idx < 0 {
+		cx.err = fmt.Errorf("uncover %d with stack size = %d", depth, len(cx.stack))
+		return
+	}
+	topIndex := len(cx.stack) - 1
+	Idxsv := cx.stack[idx]
```
Comment (jannotti) [inline data/transactions/logic/eval.go:1730]: Use lower case for locals

**Review by jannotti — COMMENTED** [review]
I think the shifts can be done clearer and faster with copy()

File: data/transactions/logic/eval.go (hunk 2 above)
Comment (AyAggarwal) [reply data/transactions/logic/eval.go:1730]: done

**Comment by AyAggarwal** [comment]
updated to use copy

**Comment by jannotti** [comment]
I'm realizing there is a tricky issue with these opcodes - normally our assembler and the Check() call try to keep track of the types on the stack. If you push an int, then a bytes, then use +, you will get an error. (In the presence of branches, we downgrade this error to a warning).

This can only happen because each opcode (so far) only manipulates the top of the stack. The OpSpec describes that manipulation in terms of types. But these opcodes are different. They manipulate the stack deeply, and potentially change the types of many elements (the element doesn't change, but the type at a particular location does, as they all shift).

We need to figure out how that can be made to work. It might just be that they need their own check function (like the branches have) or it might be that they can explicitly manipulate the type stack during assembly. They might need both, or even more.

First step, let's get a unit test in that shows the problem. Something like:

`int 4; byte "john"; int 5; cover 2; pop; +`

I think this will not type check, but it is correct (if I haven't messed up).

File: data/transactions/logic/eval.go [hunk 3]
```diff
@@ -1700,6 +1700,36 @@ func opDig(cx *evalContext) {
 	cx.stack = append(cx.stack, sv)
 }
 
+func opCover(cx *evalContext) {
+	depth := int(uint(cx.program[cx.pc+1]))
```
Comment (jdtzmn) [inline data/transactions/logic/eval.go:1704]: Why use multiple type casts?

File: data/transactions/logic/eval.go [hunk 4]
```diff
@@ -1700,6 +1700,36 @@ func opDig(cx *evalContext) {
...
+	idx := len(cx.stack) - 1 - depth
+	// Need to check stack size explicitly here because checkArgs() doesn't understand cover
+	// so we can't expect our stack to be prechecked.
+	if idx < 0 {
+		cx.err = fmt.Errorf("cover %d with stack size = %d", depth, len(cx.stack))
+		return
+	}
+	topidx := len(cx.stack) - 1
```
Comment (algochoi) [inline data/transactions/logic/eval.go:1712]: nit: `topIdx`

File: data/transactions/logic/eval.go [hunk 5]
```diff
@@ -1700,6 +1700,36 @@ func opDig(cx *evalContext) {
...
+func opUncover(cx *evalContext) {
+	depth := int(uint(cx.program[cx.pc+1]))
+	idx := len(cx.stack) - 1 - depth
+	// Need to check stack size explicitly here because checkArgs() doesn't understand uncover
+	// so we can't expect our stack to be prechecked.
+	if idx < 0 {
+		cx.err = fmt.Errorf("uncover %d with stack size = %d", depth, len(cx.stack))
+		return
+	}
+	topidx := len(cx.stack) - 1
```
Comment (algochoi) [inline data/transactions/logic/eval.go:1727]: nit: `topIdx`

File: data/transactions/logic/eval_test.go [hunk 6]
```diff
@@ -4134,6 +4134,22 @@ func TestDig(t *testing.T) {
 	testPanics(t, "int 3; int 2; int 1; dig 11; int 2; ==; return", 3)
 }
 
+func TestCover(t *testing.T) {
```
Comment (algochoi) [inline data/transactions/logic/eval_test.go:4137]: I think you *might* have to run `make sanity` on the root folder (`go-algorand`) if you haven't done so already -- I took a quick look at the Travis error and apparently there might be formatting issues.

Comment (jdtzmn) [inline data/transactions/logic/eval_test.go:4137]: It would be nice to see test cases here that check that other elements are shifted correctly

**Review by jdtzmn — CHANGES_REQUESTED** [review]
Mostly good, nice work on this! I left a few comments.

File: data/transactions/logic/eval.go [hunk 7]
```diff
@@ -1700,6 +1700,36 @@ func opDig(cx *evalContext) {
...
+	idxsv := cx.stack[idx]
```
Comment (algochoi) [inline data/transactions/logic/eval.go:1728]: nit: I think this should be `sv` (based on the `dig` opcode)

**Review by algochoi — COMMENTED** [review]
Good work! Just very minor comments. 

Based on the Travis error, maybe try running `make sanity` on the root folder or `go fmt` again?

File: data/transactions/logic/eval_test.go (hunk 6 above)
Comment (algochoi) [reply data/transactions/logic/eval_test.go:4137]: It looks like the exact statements are: 
```
  gofmt -w /home/travis/build/[secure]/go-[secure]/data/transactions/logic/eval.go
  gofmt -w /home/travis/build/[secure]/go-[secure]/data/transactions/logic/evalStateful_test.go
```
but I think `make sanity` will do the same thing?

File: data/transactions/logic/eval.go [hunk 8]
```diff
@@ -1700,6 +1700,36 @@ func opDig(cx *evalContext) {
...
+	depth := int(uint(cx.program[cx.pc+1]))
+	idx := len(cx.stack) - 1 - depth
+	// Need to check stack size explicitly here because checkArgs() doesn't understand cover
+	// so we can't expect our stack to be prechecked.
+	if idx < 0 {
+		cx.err = fmt.Errorf("cover %d with stack size = %d", depth, len(cx.stack))
+		return
+	}
+	topidx := len(cx.stack) - 1
+	topsv := cx.stack[topidx]
```
Comment (algochoi) [inline data/transactions/logic/eval.go:1713]: nit: maybe `sv` (based on the `dig` opcode)?

File: data/transactions/logic/eval_test.go (hunk 6 above)
Comment (AyAggarwal) [reply data/transactions/logic/eval_test.go:4137]: ok I just did that, hopefully it passes

Comment (AyAggarwal) [reply data/transactions/logic/eval_test.go:4137]: will add these

File: data/transactions/logic/eval.go (hunk 3 above)
Comment (AyAggarwal) [reply data/transactions/logic/eval.go:1704]: copied this from dig, but I'll dig into it

**Comment by jannotti** [comment]
It looks to me like all we need to do to address the type checking is to manipulate the type stack in the assemble function for cover/uncover. Currently they use asmDefault. So they'll need custom assemblers, but it should pretty much just call the default assembler, and manipulate the stack.

File: data/transactions/logic/eval.go (hunk 3 above)
Comment (AyAggarwal) [reply data/transactions/logic/eval.go:1704]: seems like it creates a type mismatch error if its not casted back to an int. without uint it becomes negative

Comment (jdtzmn) [reply data/transactions/logic/eval.go:1703]: Aren't uint64 immediates required to be non-negative?

**Comment by jannotti** [comment]
They are byte immediates.

Still, it seems as though this double cast is unneeded.  I recall seeing it
before, messing with it a bit, and for some reason leaving it.  I can't see
why, right now.  In C, I guess something similar would be needed, as char
is not defined to be signed or unsigned.  But byte is
definitely unsigned in Go, and the simple cast to int should preserve the
sign because there's room in int for the entire range of byte.

File: data/transactions/logic/assembler.go [hunk 9]
```diff
@@ -1101,6 +1101,63 @@ func typeDig(ops *OpStream, args []string) (StackTypes, StackTypes) {
...
+	}
+	idx := len(ops.typeStack) - depth
+	if idx >= 0 {
+		sv := ops.typeStack[idx]
+		for i := idx + 1; i < len(ops.typeStack); i++ {
+			returns[i-idx-1] = ops.typeStack[i]
+		}
+		returns[len(returns)-1] = sv
+	}
+	fmt.Println(returns)
```
Comment (jdtzmn) [inline data/transactions/logic/assembler.go:1157]: nit: `fmt.Println`

File: data/transactions/logic/assembler.go [hunk 10]
```diff
@@ -1101,6 +1101,63 @@ func typeDig(ops *OpStream, args []string) (StackTypes, StackTypes) {
...
+	anys := make(StackTypes, depth)
+	for i := range anys {
+		anys[i] = StackAny
+	}
+	returns := make(StackTypes, depth)
+	for i := range returns {
+		returns[i] = StackAny
+	}
+	idx := len(ops.typeStack) - depth
+	if idx >= 0 {
```
Comment (jdtzmn) [inline data/transactions/logic/assembler.go:1122]: Will the idx here ever be less than 0? Won't that be caught in the `opCover` function?

File: data/transactions/logic/assembler.go [hunk 11]
```diff
@@ -1101,6 +1101,63 @@ func typeDig(ops *OpStream, args []string) (StackTypes, StackTypes) {
...
+		sv := ops.typeStack[len(ops.typeStack)-1]
+		for i := idx; i < len(ops.typeStack)-1; i++ {
```
Comment (jdtzmn) [inline data/transactions/logic/assembler.go:1124]: Could this be shifted more cleanly using `copy()`?

File: data/transactions/logic/eval.go [hunk 12]
```diff
@@ -1699,6 +1699,36 @@ func opDig(cx *evalContext) {
...
+func opCover(cx *evalContext) {
+	depth := int(cx.program[cx.pc+1])
+	idx := len(cx.stack) - 1 - depth
+	// Need to check stack size explicitly here because checkArgs() doesn't understand cover
+	// so we can't expect our stack to be prechecked.
+	if idx < 0 {
+		cx.err = fmt.Errorf("cover %d with stack size = %d", depth, len(cx.stack))
+		return
+	}
+	topIdx := len(cx.stack) - 1
```
Comment (jdtzmn) [inline data/transactions/logic/eval.go:1711]: This may read better if you move `topIdx` above `idx` like so:
```go
topIdx := len(cx.stack - 1)
idx := topIdx - depth
```

**Review by jdtzmn — CHANGES_REQUESTED** [review]

File: data/transactions/logic/eval.go [hunk 13]
```diff
@@ -1699,6 +1699,36 @@ func opDig(cx *evalContext) {
 	cx.stack = append(cx.stack, sv)
 }
 
+func opCover(cx *evalContext) {
+	depth := int(cx.program[cx.pc+1])
+	idx := len(cx.stack) - 1 - depth
+	// Need to check stack size explicitly here because checkArgs() doesn't understand cover
+	// so we can't expect our stack to be prechecked.
```
Comment (jannotti) [inline data/transactions/logic/eval.go:1706]: This may not be true any more with the changes to checkStack.  I think that now opCover _is_ sure to have a big enough stack to work with.

File: data/transactions/logic/eval_test.go [hunk 14]
```diff
@@ -4204,6 +4204,25 @@ func TestDig(t *testing.T) {
...
+}
+
+func TestUncover(t *testing.T) {
+	t.Parallel()
+	testAccepts(t, "int 4; int 3; int 2; int 1; uncover 2; int 3; ==; return", 5)
+	testAccepts(t, "int 4; int 3; int 2; int 1; uncover 3; int 4; ==; return", 5)
+	testAccepts(t, "int 4; int 3; int 2; int 1; uncover 3; pop; int 1; ==; return", 5)
+	testAccepts(t, "int 4; int 3; int 2; int 1; uncover 3; pop; pop; int 2; ==; return", 5)
+	testAccepts(t, "int 1; int 3; int 2; int 1; uncover 3; pop; pop; int 2; ==; return", 5)
+	testPanics(t, obfuscate("int 4; int 3; int 2; int 1; uncover 11; int 3; ==; return"), 5)
```
Comment (jannotti) [inline data/transactions/logic/eval_test.go:4223]: Can you add one that panics, but just barely? In other words, confirm theres no off-by-one error by uncovering with an argument that's just one too high.

**Review by jannotti — APPROVED** [review]

File: data/transactions/logic/assembler.go (hunk 11 above)
Comment (jannotti) [reply data/transactions/logic/assembler.go:1124]: It could, but I'm to blame for setting the precedent with `dig`. We need cleaner helper functions for building these arrays in the type* functions.  Then we can also do `select`, `dup`, `dup`.

Comment (jdtzmn) [reply data/transactions/logic/assembler.go:1124]: Agreed. Does it make sense to include those changes in this PR or should it be done in a separate one?

Comment (jannotti) [reply data/transactions/logic/assembler.go:1124]: I think we should do those separately.  I'm still mulling over the right interface, but I think I'm imagining two things, one I'm fairly sure about, and one less so.

Fairly sure: the typeCheck functions should check the appropriate depth first, and return early if the stack is short (with an appropriately sized slice to cause checkStack to complain about length).  This means the rest of the typeCheck function can assume the stack is deep enough, and it'll be easier to write the code (using copy!) that returns slices that match parts of the existing stack.

Less sure: the interface should be reworked a little so that immediates are parsed early in one place, and supplied in a structured way to the typeCheck and assemble functions.  That way it only happens once, and assemble functions are no longer expected to do that parsing or error checking.

Comment (AyAggarwal) [reply data/transactions/logic/assembler.go:1124]: I agree with the first one for sure, the second point is good as well - would that still mean if a program failed in typeCheck it fails during assembly?

File: data/transactions/logic/eval.go (hunk 12 above)
Comment (AyAggarwal) [reply data/transactions/logic/eval.go:1711]: good point

File: data/transactions/logic/eval.go (hunk 13 above)
Comment (AyAggarwal) [reply data/transactions/logic/eval.go:1706]: good point, i'll re visit this

File: data/transactions/logic/assembler.go (hunk 10 above)
Comment (AyAggarwal) [reply data/transactions/logic/assembler.go:1122]: I believe this check occurs first, as jj said below opCover should now be sure that it has the correct size stack

Comment (AyAggarwal) [reply data/transactions/logic/assembler.go:1122]: This is actually because the arrays will create out of bounds error if you don't check here. It doesn't panic though

**Review by jdtzmn — APPROVED** [review]

File: data/transactions/logic/doc.go [hunk 15]
```diff
@@ -246,7 +250,7 @@ var OpGroups = map[string][]string{
 	"Byteslice Arithmetic": {"b+", "b-", "b/", "b*", "b<", "b>", "b<=", "b>=", "b==", "b!=", "b%"},
 	"Byteslice Logic":      {"b|", "b&", "b^", "b~"},
 	"Loading Values":       {"intcblock", "intc", "intc_0", "intc_1", "intc_2", "intc_3", "pushint", "bytecblock", "bytec", "bytec_0", "bytec_1", "bytec_2", "bytec_3", "pushbytes", "bzero", "arg", "arg_0", "arg_1", "arg_2", "arg_3", "txn", "gtxn", "txna", "gtxna", "gtxns", "gtxnsa", "global", "load", "store", "gload", "gloads", "gaid", "gaids"},
-	"Flow Control":         {"err", "bnz", "bz", "b", "return", "pop", "dup", "dup2", "dig", "swap", "select", "assert", "callsub", "retsub"},
+	"Flow Control":         {"err", "bnz", "bz", "b", "return", "pop", "dup", "dup2", "dig", "cover", "uncover", "swap", "select", "assert", "callsub", "retsub"},
```
Comment (algorandskiy) [inline data/transactions/logic/doc.go:253]: just wondering how did `"pop", "dup", "dup2", "dig", "cover", "uncover"` end up to be flow control

File: data/transactions/logic/eval_test.go [hunk 16]
```diff
@@ -4400,6 +4400,27 @@ func TestDig(t *testing.T) {
...
+	testAccepts(t, "int 4; int 3; int 2; int 1; cover 1; int 2; ==; return", 5)
+	testAccepts(t, "int 4; int 3; int 2; int 1; cover 2; int 2; ==; return", 5)
+	testAccepts(t, "int 4; int 3; int 2; int 1; cover 2; pop; pop; int 1; ==; return", 5)
+	testPanics(t, obfuscate("int 4; int 3; int 2; int 1; cover 11; int 2; ==; return"), 5)
+	testPanics(t, obfuscate("int 4; int 3; int 2; int 1; cover 4; int 2; ==; return"), 5)
+}
+
+func TestUncover(t *testing.T) {
+	t.Parallel()
+	testAccepts(t, "int 4; int 3; int 2; int 1; uncover 2; int 3; ==; return", 5)
```
Comment (algorandskiy) [inline data/transactions/logic/eval_test.go:4415]: please add `unvcover 0` test as an edge case test

File: data/transactions/logic/eval.go [hunk 17]
```diff
@@ -1699,6 +1699,36 @@ func opDig(cx *evalContext) {
...
+func opUncover(cx *evalContext) {
+	depth := int(cx.program[cx.pc+1])
+	idx := len(cx.stack) - 1 - depth
+	// Need to check stack size explicitly here because checkArgs() doesn't understand uncover
+	// so we can't expect our stack to be prechecked.
+	if idx < 0 {
+		cx.err = fmt.Errorf("uncover %d with stack size = %d", depth, len(cx.stack))
+		return
+	}
+	topIdx := len(cx.stack) - 1
```
Comment (algorandskiy) [inline data/transactions/logic/eval.go:1726]: make it the same as prev function
```suggestion
	depth := int(cx.program[cx.pc+1])
	topIdx := len(cx.stack) - 1
	idx := topIdx - depth
	// Need to check stack size explicitly here because checkArgs() doesn't understand uncover
	// so we can't expect our stack to be prechecked.
	if idx < 0 {
		cx.err = fmt.Errorf("uncover %d with stack size = %d", depth, len(cx.stack))
		return
	}
```

**Review by algorandskiy — COMMENTED** [review]
Looks good, one more test request and a minor code style fix

File: data/transactions/logic/doc.go (hunk 15 above)
Comment (jannotti) [reply data/transactions/logic/doc.go:253]: I think it's because it used to only be `pop` and `dup` so we didn't create a category for stack manipulation.

**Review by jannotti — APPROVED** [review]
Looks good.
